### DISCUSSION TITLE
Return result to back navigation

### DISCUSSION THREAD

**Discussion by JajaComp** [arkivanov/Decompose#678] [Q&A]
Hi! I try create common input text dialog and return result to called component.
InputDialog:
```
 InputDialogComponentImpl(
    private val _inputResult = mutableResultFlow<String>()
    val inputResult = _inputResult.asSharedFlow()
    ...
    navResult = { result ->
        navigation.dismiss {
            _inputResult.tryEmit(...)
        }
    },
),
```
RootComponent:
```
class RootComponentImpl{
    init {
        coroutineScope().launch {
          repeatOnLifecycle(
              minActiveState = Lifecycle.State.RESUMED,
          ) {
              dialogComponent.inputResult.collect {
                  println(it)
              }
          }
    }
}
```
But I don’t get the result because when the message arrives, the component is not yet in the resumed state. How i can fix it? Mb replace _inputResult.tryEmit to _inputResult.emit?

**Comment by JajaComp** [comment]
Now my best solution is:
```
    private fun <T> MutableSharedFlow<T>.pushResult(value: T) {
        resetResult()
        resultJob = scope.launch {
            this@pushResult.subscriptionCount.first { it > 0 } // wait first subscriber
            this@pushResult.emit(value)
        }
    }

    private fun resetResult() {
        resultJob?.cancel()
        resultJob = null
    }
```
resetResult also called when call childFactory

**Comment by arkivanov** [comment]
I think you can just pass a callback to your `InputDialogComponentImpl` class. You can also check the [docs](https://arkivanov.github.io/Decompose/navigation/stack/overview/#delivering-a-result-when-navigating-back).

**Reply by JajaComp** [reply]
The callback solution will not work in this case because InputDialogComponentImpl can be used by any other component

**Reply by arkivanov** [reply]
Based on the provided code, it's not clear what the actual case this is. But you can also try to avoid repeatOnLifecycle and just always collect in the scope.

**Reply by JajaComp** [reply]
This sample of situation when user can input text using dialog in few places of application. If i will avoid repeatOnLifecycle then i will got issue in this situation:
Navigation stack:
- ComponentA
- ComponentB -> collect of inputResultFlow
- ComponentC -> collect of inputResultFlow
When ComponentC will got response from input, ComponentB also will got same result.
Can i call onComplete callback of pop/dismiss method after change status of parent component to Resumed lifecycle status?

**Reply by arkivanov** [reply]
Based on what you described, I would just host the dialog component in every component where required. One of the way is to use [Child Slot](https://arkivanov.github.io/Decompose/navigation/slot/overview/) navigation model in every component where you need the dialog. And still use a callback for delivering results. Would this work for you?
